keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,343 @@
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non\-Goals](#non-goals)
+  * [User Stories](#user-stories)
+* [Proposal](#proposal)
+  * [API](#api)
+    * [Key Secret Definition](#key-Secret-definition)
+    * [Image Handler](#Image-handler)
+    * [Container Handler](#Container-handler)
+  * [Relationship with imagePullPolicy](#Relationship-with-imagePullPolicy)
+  * [Consumption of the ImageDecryptSecrets](#Consumption-of-the-ImageDecryptSecrets)
+* [Alternatives Considered](#alternatives-considered)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The underlying specification for the containers, the OCI spec, is soon going to support encrypted images. Kubernetes should be able to support decryption of these encrypted images with the addition of a new type of `Secret`, which we would like to call `ImageDecryptSecret`. 
+
+Along with OCI spec, there is an ongoing effort to enable the support for encrypted images in containerd. 
+
+OCI Spec Issue - https://github.com/opencontainers/image-spec/issues/747
+
+OCI Spec PR - https://github.com/opencontainers/image-spec/pull/775
+
+Containerd StreamProcessors PR - https://github.com/containerd/containerd/pull/3482 and the container image decryption library that works with it, https://github.com/stefanberger/imgcrypt/
+
+POC - https://github.com/harche/kubernetes/tree/pr_branch
+
+Suported CRI-O implementation - https://github.com/harche/cri-o/tree/pull_img_auth 
+
+
+## Motivation
+
+The kubernetes worker nodes are where container images are pulled by a runtime such as `containerd`. If the images pulled are encrypted then containerd will have to decrypt them before running the pod. In order to be able to decrypt the images, the worker node needs to have access to corresponding the private keys.
+
+Kubernetes `Secrets` are used to securely deliver sensitive data to pods in corresponding worker nodes. We need to have a secret that can be utilized *before* the pod is provisioned. Regular Kubernetes secret gets attached to the pod as tmpfs mount after the pod is started. However, there exists another type of kubernetes secret called `ImagePullSecrets`. ImagePullSecrets are used to pull the images from the private container image registry, hence they contain login credentials. `ImagePullSecrets` get utilized *before* the pod is started, this is exactly the same kind of requirement for being able to decrypt the encrypted container image. We need to have a `secret` that can be used *before* the pod is provisioned to decrypt the image. Hence, we are submitting this KEP to kubernetes community to propose a new type of kubernetes secret that is modeled after `ImagePullSecret`, called `ImageDecryptSecret`. While the `ImagePullSecret` holds the login credentials for the private registry, the `ImageDecryptSecret` will hold the private keys to decrypt encrypted images.
+
+
+### Goals
```

Comment:
I'd like to see the KEP deal with why this must be a field on pods, vs implemented by container runtimes and out of scope for Kubernetes.  Specifically, discussion of the threat model that this KEP is working against where secrets (visible to any administrator of the cluster) are being used to decrypt images, vs node level secrets that the container runtime uses to decrypt the contents.

I'd also like to see some ecosystem adoption (OCI + at least one container runtime) before we move this forward to Kubernetes.  I don't think 1.17 is an appropriate milestone for an in progress work (although I welcome the KEP and PRS or feature branches that demonstrate this behavior).

The KEP also needs to discuss how container runtimes that opt not to implement this extension would signal that, how errors propagated to the Kubelet would be reflected all the way back to the user interface, and how a conformant Kubernetes cluster would react if this capability is not present.

Reply:
Hi @smarterclayton, thanks for your feedback.

I have added the following sections to talk about,
1. [Compatibility with the runtime](https://github.com/kubernetes/enhancements/blob/899bd65f58c13288faa11b0451efaef15a6f2b27/keps/sig-node/20190517-image-decryption.md#Runtime-compatibility)
2. [Threat Model ](https://github.com/kubernetes/enhancements/blob/899bd65f58c13288faa11b0451efaef15a6f2b27/keps/sig-node/20190517-image-decryption.md#Threat-model)